File: ECnotes/Sem1/Labs/LivingExamples/VB/Ch16CategoryMaint/App_Code/CategoryDB.vb
Imports Microsoft.VisualBasic
Imports System.Data
Imports System.Data.SqlClient
Imports System.ComponentModel

<DataObject(True)> _
Public Class CategoryDB

    <DataObjectMethod(DataObjectMethodType.Select)> _
    Public Shared Function GetCategories() As IEnumerable
        Dim sel As String = "SELECT CategoryID, ShortName, LongName " _
            & "FROM Categories ORDER BY ShortName"
        Dim cmd As SqlCommand = _
            New SqlCommand(sel, New SqlConnection(GetConnectionString))
        cmd.Connection.Open()
        Return cmd.ExecuteReader(CommandBehavior.CloseConnection)
    End Function

    Private Shared Function GetConnectionString() As String
        Return ConfigurationManager.ConnectionStrings _
            ("HalloweenConnection").ConnectionString
    End Function

    <DataObjectMethod(DataObjectMethodType.Insert)> _
    Public Shared Sub InsertCategory(ByVal CategoryID As String, _
            ByVal ShortName As String, ByVal LongName As String)
        Dim ins As String = "INSERT INTO Categories " _
            & "(CategoryID, ShortName, LongName) " _
            & "VALUES(@CategoryID, @ShortName, @LongName)"
        Dim cmd As SqlCommand = _
            New SqlCommand(ins, New SqlConnection(GetConnectionString))
        cmd.Parameters.AddWithValue("CategoryID", CategoryID)
        cmd.Parameters.AddWithValue("ShortName", ShortName)
        cmd.Parameters.AddWithValue("LongName", LongName)
        cmd.Connection.Open()
        cmd.ExecuteNonQuery()
    End Sub

    <DataObjectMethod(DataObjectMethodType.Delete)> _
    Public Shared Function DeleteCategory( _
            ByVal original_CategoryID As String, _
            ByVal original_ShortName As String, _
            ByVal original_LongName As String) _
            As Integer
        Dim del As String = "DELETE FROM Categories " _
            & "WHERE CategoryID = @original_CategoryID " _
            & "AND ShortName = @original_ShortName " _
            & "AND LongName = @original_LongName "
        Dim cmd As SqlCommand = _
            New SqlCommand(del, New SqlConnection(GetConnectionString))
        cmd.Parameters.AddWithValue("original_CategoryID", original_CategoryID)
        cmd.Parameters.AddWithValue("original_ShortName", original_ShortName)
        cmd.Parameters.AddWithValue("original_LongName", original_LongName)
        cmd.Connection.Open()
        Dim i As Integer = cmd.ExecuteNonQuery()
        cmd.Connection.Close()
        Return i
    End Function

    <DataObjectMethod(DataObjectMethodType.Update)> _
    Public Shared Function UpdateCategory(ByVal ShortName As String, _
            ByVal LongName As String, ByVal original_CategoryID As String, _
            ByVal original_ShortName As String, _
            ByVal original_LongName As String) As Integer
        Dim up As String = "UPDATE Categories " _
            & "SET ShortName = @ShortName, " _
            & "LongName = @LongName " _
            & "WHERE CategoryID = @original_CategoryID " _
            & "AND ShortName = @original_ShortName " _
            & "AND LongName = @original_LongName"
        Dim cmd As SqlCommand = _
            New SqlCommand(up, New SqlConnection(GetConnectionString))
        cmd.Parameters.AddWithValue("ShortName", ShortName)
        cmd.Parameters.AddWithValue("LongName", LongName)
        cmd.Parameters.AddWithValue("original_CategoryID", original_CategoryID)
        cmd.Parameters.AddWithValue("original_ShortName", original_ShortName)
        cmd.Parameters.AddWithValue("Original_LongName", original_LongName)
        cmd.Connection.Open()
        Dim i As Integer = cmd.ExecuteNonQuery()
        cmd.Connection.Close()
        Return i
    End Function

End Class

File: ECnotes/Sem1/Labs/LivingExamples/CS/Ch16CategoryMaint/App_Code/CategoryDB.cs
using System;
using System.Collections;
using System.Data;
using System.Data.SqlClient;
using System.Configuration;
using System.ComponentModel;

[DataObject(true)]
public static class CategoryDB
{
    [DataObjectMethod(DataObjectMethodType.Select)]
    public static IEnumerable GetCategories()
    {
        string sel = "SELECT CategoryID, ShortName, LongName "
                   + "FROM Categories ORDER BY ShortName";

        SqlCommand cmd = new SqlCommand(sel, new SqlConnection(GetConnectionString()));

        cmd.Connection.Open();

        SqlDataReader dr = cmd.ExecuteReader(CommandBehavior.CloseConnection);

        return dr;
    }

    private static string GetConnectionString()
    {
        return ConfigurationManager.ConnectionStrings["HalloweenConnectionString"].ConnectionString;
    }

    [DataObjectMethod(DataObjectMethodType.Insert)]
    public static int InsertCategory(string CategoryID, string ShortName, string LongName)
    {
        string ins = "INSERT INTO Categories "
                    + " (CategoryID, ShortName, LongName) "
                    + " VALUES(@CategoryID, @ShortName, @LongName)";

        SqlCommand cmd = new SqlCommand(ins, new SqlConnection(GetConnectionString()));

        cmd.Parameters.AddWithValue("CategoryID", CategoryID);
        cmd.Parameters.AddWithValue("ShortName", ShortName);
        cmd.Parameters.AddWithValue("LongName", LongName);

        cmd.Connection.Open();

        int i = cmd.ExecuteNonQuery();

        return i;
    }

    [DataObjectMethod(DataObjectMethodType.Delete)]
    public static int DeleteCategory(string original_CategoryID,
        string original_ShortName, string original_LongName)
    {
        string del = "DELETE FROM Categories "
                   + "WHERE CategoryID = @original_CategoryID "
                   + "  AND ShortName = @original_ShortName "
                   + "  AND LongName = @original_LongName ";

        SqlCommand cmd = new SqlCommand(del, new SqlConnection(GetConnectionString()));

        cmd.Parameters.AddWithValue("original_CategoryID", original_CategoryID);
        cmd.Parameters.AddWithValue("original_ShortName", original_ShortName);
        cmd.Parameters.AddWithValue("original_LongName", original_LongName);
        cmd.Connection.Open();

        int i = cmd.ExecuteNonQuery();

        return i;
    }

    [DataObjectMethod(DataObjectMethodType.Update)]
    public static int UpdateCategory(string ShortName, 
        string LongName, string original_CategoryID, 
        string original_ShortName, string original_LongName)
    {
        string up = "UPDATE Categories "
                  + "SET ShortName = @ShortName, "
                  + "LongName = @LongName "
                  + "WHERE CategoryID = @original_CategoryID "
                  + "  AND ShortName = @original_ShortName "
                  + "  AND LongName = @original_LongName";

        SqlCommand cmd = new SqlCommand(up, new SqlConnection(GetConnectionString()));

        cmd.Parameters.AddWithValue("ShortName", ShortName);
        cmd.Parameters.AddWithValue("LongName", LongName);
        cmd.Parameters.AddWithValue("original_CategoryID", original_CategoryID);
        cmd.Parameters.AddWithValue("original_ShortName", original_ShortName);
        cmd.Parameters.AddWithValue("original_LongName", original_LongName);

        cmd.Connection.Open();

        int i = cmd.ExecuteNonQuery();

        return i;
    }
}
